cdhweb/projects/migrations/0011_calculate_membership_dates.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.23 on 2020-11-10 17:10
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.core.exceptions import ObjectDoesNotExist
+
+
+def calculate_membership_dates(apps, schema_editor):
+    # We can't import the Person model directly as it may be a newer
+    # version than this migration expects. We use the historical version.
+    Membership = apps.get_model('projects', 'Membership')
+    # sort by project, then user, then grant start date (role?)
+    memberships = Membership.objects.all() \
+                            .order_by('project', 'user', 'grant__start_date')
+
+    # as we loop through, track current membership to extend across grants
+    current = None
+    for membership in memberships:
+        if current:
+            # if this membership matches the previous one, extend it
+            if current.user == membership.user and \
+               current.project == membership.project and \
+               current.role == membership.role:
+                # extend the end date
+                current.end_date = calculate_membership_end(membership)
+                # delete the obsolete membership
+                membership.delete()
+                continue
```

Comment:
there is theoretically a scenario here where a person (maybe a PM?) is involved in the same role with a project that gets two overlapping grants, one of which is longer (a year, say) and the other of which is shorter (a few months) and concludes before the longer one ends. in this case, the end date can get updated to point at the end of the second grant, even though the first one lasts longer.

i have no idea if this would ever happen in practice, but in any case I think you can mitigate it by just adding a check here before extending:
```py
calculate_membership_end(current) < calculate_membership_end(membership)
```

Reply:
Hm. Any idea if this is a real scenario? I don't think our grant dates typically overlap.

I can use a maximum, but the logic is slightly more complicated because in the current logic `None` could be a valid value for an end date if the status override is set to current.

Reply:
It's probably not real, so if it's not trivial to handle let's not worry about it.